### DISCUSSION TITLE
Three questions (Scrolling inertia, perspective, control template)

### DISCUSSION THREAD

**Discussion by DarthMazut** [AvaloniaUI/Avalonia#13465] [Q&A]
Hello, I have couple of questions, and I thought it will be good idea to ask all of them in one post:
1. Is it possible to achieve scrolling inertia effect on ScrollViewer? I already set `IsScrollInertiaEnabled` to `true` on my ScrollViewer but it didn't work for some reason? My XAML looks like this:
``` xml
<ScrollViewer x:Name="PART_ScrollViewer"
			  Offset="{TemplateBinding Offset, Mode=TwoWay}"
			  HorizontalScrollBarVisibility="Hidden"
			  VerticalScrollBarVisibility="Hidden"
			  IsScrollInertiaEnabled="True">
```
and code for panning look like that:
```cs
        public void Pan(PointerEventArgs e)
        {
            Point currentPoinerPosition = e.GetPosition(_scrollViewer);
            Vector newOffset = Offset - (currentPoinerPosition - _lastPointerPosition);
            newOffset = new Vector(
                Math.Min(Math.Max(0, newOffset.X), _scrollViewer.ScrollBarMaximum.X),
                Math.Min(Math.Max(0, newOffset.Y), _scrollViewer.ScrollBarMaximum.Y));

            Offset = newOffset;
        }
```
currently it works as expected but I'd like to add scrolling inertia effects, is it possible by just manipulating Offset, or some additional work is required?

2. Is it possible in AvaloniaUI to add perspective to controls? Let's say Image control.
What I already tried is to set `MatrixTransform` on the control, but when I'm manipulating `perspX`, `perspY`, `perspZ` values of `Matrix` it looks very odd, and I tried many different values there. What I'd like to achieve is something like this:
![obraz](https://github.com/AvaloniaUI/Avalonia/assets/29342390/a935795e-d7a5-4bc8-94e3-2235473f4bb6)
 
3. My last question is about the difference in designing a UserControl. Which of these approach is better and how they differ?
a)
```xml
<UserControl ...>
	<UserControl.Template>
		<ControlTemplate>
			<!-- CONTENT GOES HERE -->
		</ControlTemplate>
	</UserControl.Template>
</UserControl>
```
vs
b)
```xml
<UserControl ...>
	<!-- CONTENT GOES HERE -->
</UserControl>
```
So I noticed that in first example I'm able to use `TemplateBinding`, but is this recommend approach or somehow better than second one? If I'm not mistaken the option a) allows to display content of the UserControl by {TemplateBinding Content}, while this is not possible through binding {Binding Content}. But are there any other differences? Which should be preferred by default?

Thank you for all responses 🙂

**Comment by thevortexcloud** [comment]
Not sure about your first two questions, but for the last one both approaches can be functionally the same. You typically set the template of a control if you need to override it's appearance, you are making a theme (EG t[ake a look at the Fluent theme code in the repo](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/CheckBox.xaml)), or you are making a control that does not directly bind to a DataContext (EG any of the built in controls). There are some special rules with control names that need to be followed if you are overriding a built in control though. In any case, the first approach well let you bind to the Styled/Direct properties of the control itself using `TemplateBinding`. EG you can set a border thickness by defining a border thickness property in the control and then bind the border to that. There are some examples [here](https://docs.avaloniaui.net/docs/authoring-controls/defining-properties) showing how to declare them.

The second approach is far simpler if you want to bind to a data context that is set on the control. You can still bind to the control itself, albeit it's far less direct to do so. EG you can use `RelativeSource` bindings. It's better to do it that way if you are using MVVM and are not making very generic controls.

**Comment by timunie** [comment]
> Is it possible to achieve scrolling inertia effect on ScrollViewer? I already set IsScrollInertiaEnabled to true on my ScrollViewer but it didn't work for some reason? 

This is only implemented for mobile devices / touch screen for now. I guess it is a similar question to #12945 ? 

> Is it possible in AvaloniaUI to add perspective to controls? Let's say Image control.

Maybe try: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Media/SkewTransform.cs

> My last question is about the difference in designing a UserControl. Which of these approach is better and how they differ?

Use typically options b) as overriding ControlTemplate is normally done for re-usable custom controls. Maybe you want to have differnet DataTemplates, which is a different story.

**Reply by dbriard** [reply]
About 2) You should be able to do what you want using Rotate3DTransform.

**Reply by DarthMazut** [reply]
Thank you, it works. I had to provide `AngleX` and quite large value for `Depth`. Though I'm still facing some issues while putting my control with 3DTransform within ScrollViewer. Instead of scrolling left-right it feels more like looking around. I'm not sure whether that is normal with 3D transformation or I messed something with my layout
